### ISSUE TITLE
[improvement/feature] Local opam configuration in the current directory

### ISSUE THREAD

**Issue by dannywillems** [ocaml/opam#2545]
Node Package Manager ([NPM](https://www.npmjs.com/)) has two different possible configurations: locally and globally.

By globally, I mean you can install packages globally and use these packages in all your projects, wherever the current directory you are in. You can use `npm install -g <package>` to instal globally.

By locally, I mean you can install specific packages for your project in the current directory. Sometimes, you don't want to have `package1` in your `project1` for some compatibility reasons with `package2` in your `project2` or you want to have a specific version of `package1` for your `project1`, different for any other projects.
You can also have a `package.json` file which lists dependencies and you can install them **locally** with `npm install`.

With opam, you can only install globally. You can't install locally like NPM allows.

What you can do is change your compiler and install with opam, but it stills globally. It's a way to have "local" configuration, but it's very boring.

Is there a way to do something like NPM? If not, I think it could be a great feature to implement.

**Comment by seliopou** [comment]
Related branches can be found [here](https://github.com/inhabitedtype/opam/tree/package) and [here](https://github.com/inhabitedtype/opam/commits/local), and are built on top of next. Pull requests coming next week.

**Comment by AltGr** [comment]
@seliopou just merged next, so you can now PR them directly on master ☺

**Comment by dannywillems** [comment]
Will it be in OPAM 2.0?

**Comment by AltGr** [comment]
A similar workflow will be supported, all the atoms are here so that it is already possible -- defining the exact workflow and adding some sugar so that it becomes _convenient_ is on its way.
See the [announcement](https://opam.ocaml.org/blog/opam-2-0-preview/)
and an example in the wiki page (feel free to extend or add other examples): https://github.com/ocaml/opam/wiki/opam-2.0-FAQ

**Comment by dannywillems** [comment]
@AltGr Thanks for this answer and for the work of everyone. I just switched on 2.0.

Is there a reason `opam switch create <directory> <compiler>` creates `_opam` and not `.opam` in the current directory?

**Comment by AltGr** [comment]
No strong reason, `.opam` would work as well, but some light ones:
- avoid confusion with `~/.opam`, or don't give the impression that it holds similar contents (`_opam` being just a switch prefix and not a full opam root)
- similar to `_build` and `_obuild`, which hold build artefacts, and that seems to match, more or less
- no need for a hidden directory

**Comment by dannywillems** [comment]
* If I understand:
```
opam switch create switch_name 4.03.0
```
creates a new global switch of 4.03.0 but
```
opam switch create switch_name/ 4.03.0
```
creates a new local switch of 4.03.0 in the `switch_name/`directory (and creates the directory if not exists).
I think it's not very intuitive and error prone.
I thought about an option like `--local` (so `opam switch create --local <directory> <compiler>`) and `--global` (so `opam switch create --global <switch_name> <compiler>`).

* I have also an error if two versions of 4.03.0 are installed when I want to do a new switch to 4.03.0 -->
```
opam switch create test 4.03.0
```
outputs
```
[ERROR] Compiler selection '4.03.0' is ambiguous. matching packages: { ocaml-base-compiler.4.03.0, ocaml-system.4.03.0 }
```
and I need to write
```
opam switch create test ocaml-base-compiler.4.03.0
```

**Comment by dannywillems** [comment]
@AltGr I see you're closing a lot of issue. I think this one can be closed as it's implemented in OPAM 2.0. Maybe just read my last comment.

**Comment by AltGr** [comment]
Yes, trying to clean up before releasing the 2.0 beta.
Thanks for the feedback. I am afraid the latter error can't be avoided, actually `opam switch create 4.03.0` is here not to break habits, but is a somewhat magical shortcut (that attempts to match on compiler package versions rather than names). The longer version is the explicit syntax.
